File: merge_sort/Python/merge_sort.py
"""
Merge Sort Algorthm
- Create an empty list called the result list.
- Do the following until one of the input lists is empty:
    - Remove the first element of the list that has a lesser first element
      and append it to the result list.
- When one of the lists is empty, append all elements of the other list to
  the result.
"""


def _merge(left, right):
    result = []
    left_idx, right_idx = 0, 0
    while left_idx < len(left) and right_idx < len(right):
        # change the direction of this comparison to change
        # the direction of the sort
        if left[left_idx] <= right[right_idx]:
            result.append(left[left_idx])
            left_idx += 1
        else:
            result.append(right[right_idx])
            right_idx += 1

    if left:
        result.extend(left[left_idx:])
    if right:
        result.extend(right[right_idx:])
    return result


def merge_sort(arr):
    """
    Fuction to do merge sort.

    :param arr: A list of element to sort.
    """
    if len(arr) <= 1:
        return arr

    middle = len(arr) // 2
    left = arr[:middle]
    right = arr[middle:]

    left = merge_sort(left)
    right = merge_sort(right)
    return list(_merge(left, right))


def main():
    arr = [6, 5, 4, 3, 2, 1]
    print('Sorted element using Merge Sort: {}'.format(
              ' '.join(map(str, merge_sort(arr)))))


if __name__ == '__main__':
    main()

File: Sorting/mergesort.cpp
#include <iostream>
using namespace std;


void merge(int a[], int n1, int b[], int n2, int arr[])
{
  int i=0,j=0,k=0;
  cout<<endl;
  while(i<n1)
  {
  	cout<<a[i]<<" ";
  	i++;
  }
  cout<<endl;
  while(j<n2)
  {
  	cout<<b[j]<<" ";
  	j++;
  }
i=0;j=0;
  while(i<n1&&j<n2)
  {
    if(a[i]<b[j])
    {
      arr[k]=a[i];
      i++;
    }
    else
    {
      arr[k]=b[j];
      j++;
    }
    k++;
  }
  while(i<n1)
  {
    arr[k]=a[i];
    k++;
    i++;
  }
  while(j<n2)
  {
    arr[k]=b[j];
    k++;
    j++;
  }
}
void mergeSort(int input[], int n){
	
  if(n==0||n==1)
    return;
  else
  {
    int mid=(n-1)/2;
    int n1=mid+1;
    int n2=n-mid-1;
    int a[n1],b[n2];
    for(int i=0;i<n1;i++)
      {
      	a[i]=input[i];
      }
    int k=0;
    for(int j=mid+1;j<n;j++)
    {
      b[k]=input[j];
      k++;
    }
    mergeSort(a,n1);
    mergeSort(b,n2);
    merge(a,n1,b,n2,input);
  }
        
}
int main()
{
  int n;
  cin>>n;
  int arr[n];
  for(int i=0;i<n;i++)
    cin>>arr[i];

  mergeSort(arr,n);
  for(int i=0;i<n;i++)
    cout<<arr[i]<<" ";
}


